### ISSUE TITLE
Hoe to use Mobile Drag Drop Library on the AngularJS platform

### ISSUE THREAD

**Issue by WisarutBholsithi** [timruffles/mobile-drag-drop#161]
Hi everybody, 

I have tried using Mobile Drag Drop Library to deal with Drag Drop issue on Mobile Device by following the example from the demo. So far, this library works well in my pure HTML5 - Javascript application.

However, when I tried to modify the code to be applied to  AngularJS platform, it start to show the error as follows: 
![ProblemwithMobileDragDroponAngularJS](https://user-images.githubusercontent.com/58510558/112987384-dd595400-918c-11eb-86bc-487171642bf6.jpg)

Here is the code on pure HTML5 - Javascript application:
```
<!-- Main library of Mobile Drag Drop -->
	<script src="../static/js/mobile-drag-drop/index.min.js"></script>
	<!--optional import of scroll behaviour-->
	<script src="../static/js/mobile-drag-drop/scroll-behaviour.min.js"></script>
	<!-- function to set events for components - critical -->
	<script src="../static/js/demo.js"></script>
......

	// workaround for iOS10/iOS11 touchmove behaviour (https://github.com/timruffles/mobile-drag-drop/issues/77)
    try {            
		window.addEventListener('touchmove', function () {
		}, {passive: false})
    } catch(e) {}
	
	
	var urlParams = new URLSearchParams(window.location.search);
    var holdToDrag = urlParams.get('holdToDrag');
	var appliedMessage = "Polyfill inactive: Current browser was detected to handle drag'n'drop quite well on its own ;)";
	console.log(appliedMessage);
	var polyfillApplied = MobileDragDrop.polyfill({
		dragImageTranslateOverride: MobileDragDrop.scrollBehaviourDragImageTranslateOverride,
        holdToDrag: holdToDrag
	});
	if (polyfillApplied) {

		appliedMessage = "Polyfill activated: Detected current browser not capable of HTML5 drag'n'drop (or somebody set forceApply is to true)";

        if(holdToDrag) {
                appliedMessage += ". holdToDrag is active with a delay of " + holdToDrag + "ms";
        }
    }
	console.log(appliedMessage);
	
	var ImageDirectoryElement = document.getElementById('ImageDirectory');
	ImageDirectoryElement.addEventListener("dragstart", dragStart, false);
	
	
	function dragStart(e)
	{	
		//when the drag starts
		itemURL = e.target.src;
		 
		if (e.target !== stage) {  //if user is dragging circle
			activeBool = true; //the drag is active
		}
		
	}
	// Set Event for Drag drop to Image elements!
	var ImageforSettingDrags = document.querySelectorAll('#ImageDirectory img'), el = null;
	// the way to set properties to image elements  
	for (var i = 0; i <= ImageforSettingDrags.length - 1; i++) {
        el = ImageforSettingDrags[i];
        el.setAttribute('draggable', 'true');
		addEvent(el, 'dragstart', function (e) {
            console.log('dragstart for image' + this.id);

            e.dataTransfer.effectAllowed = 'copyLink';
			console.log('setting data: ' + this.id);

            e.dataTransfer.setData('Text', this.id); // required otherwise doesn't work
        });
		addEvent(el, 'drag', function (e) {
            console.log('drag');
        });

        addEvent(el, 'dragend', function (e) {
           console.log('dragend with effect: ' + e.dataTransfer.dropEffect);
        });
		addEvent(el, 'click', function (e) {
            console.log('Image ' + this.id +' clicked on draggable');
            console.log('click offsetX, offsetY of image' + this.id + ': ' + e.offsetX + ", " + e.offsetY);
        });

        // Custom event fires only if holdToDrag is true - critical to allow drag drop on mobile device. 
        addEvent(el, 'dnd-poly-dragstart-pending', function (e) {
            console.log('dnd-poly-dragstart-pending');

            e.preventDefault();
            e.stopPropagation(); // stop it here to prevent it bubble up
        });

        addEvent(el, 'dnd-poly-dragstart-cancel', function (e) {
            console.log('dnd-poly-dragstart-cancel');
            e.preventDefault();
            e.stopPropagation(); // stop it here to prevent it bubble up
        });
			
	}
	var draggables = document.querySelectorAll('.draggable');
	for (var i = 0; i <= draggables.length -1; i++) {
        makeElementDraggable(draggables[i]);
    }
	function makeElementDraggable(element) {

        element.setAttribute('draggable', 'true');

        addEvent(element, 'dragstart', function (e) {

        // Check whether the element is draggable, since dragstart might be triggered on a child.
        if (element.getAttribute('draggable') == 'false') return true;

                console.log('dragstart');

                e.dataTransfer.effectAllowed = 'copyLink';

                console.log('setting data: ' + this.id);

                e.dataTransfer.setData('Text', this.id); // required otherwise doesn't work

                // Try setting a proper drag image if triggered on a dnd-handle (won't work in IE).
                if (e._dndHandle && e.dataTransfer.setDragImage) {

                    e.dataTransfer.setDragImage(element);
                }
        });

        addEvent(element, 'drag', function (e) {
                console.log('drag Element');
            });

            addEvent(element, 'dragend', function (e) {
                console.log('dragend Element with effect: ' + e.dataTransfer.dropEffect);
            });

        addEvent(element, 'click', function (e) {
                console.log('clicked on draggable');
        });
    }

```
Here is the code on HTML5 - AngularJS application:
```
import {Config} from 'mobile-drag-drop/index'; // link to TS file to deal with drag drop for mobile device 
import {ScrollOptions} from 'mobile-drag-drop/scroll-behaviour'; // link to TS file to deal with drag drop for mobile device 
.. 

ngOnInit(): void {
			this.templatesService.getList().pipe(take(1)).subscribe(templates => {
			console.log(templates);
			// workaround for iOS10/iOS11 touchmove behaviour (https://github.com/timruffles/mobile-drag-drop/issues/77)
			try {            
				window.addEventListener('touchmove', function () {}, {passive: false})
			} catch(e) {}
			let urlParams = new URLSearchParams(window.location.search);
			let holdToDrag = urlParams.get('holdToDrag');
			let appliedMessage = "Polyfill inactive: Current browser was detected to handle drag'n'drop quite well on its own ;)";
			console.log(appliedMessage);
			**let polyfillApplied = MobileDragDrop.polyfill({
				dragImageTranslateOverride: MobileDragDrop.scrollBehaviourDragImageTranslateOverride,
				holdToDrag: holdToDrag
			}); // Problem start here!**
			if (polyfillApplied) {

				appliedMessage = "Polyfill activated: Detected current browser not capable of HTML5 drag'n'drop (or somebody set forceApply is to true)";

				if(holdToDrag) {
					appliedMessage += ". holdToDrag is active with a delay of " + holdToDrag + "ms";
				}
			}
			console.log(appliedMessage);
	
			const imageDir: ImageDirectory[] = CATEGORY.map(category => {
				return {
					type: category,
					images: templates.filter(t => t.category === category).map((t, i) => {
						return {
							title: `${category}_${i}`,
							src: this.templatesService.get(t.fileId),
							remarks: t.remarks,
							subCategory: t.sub_category,
							category: t.category
						};
					})
				} as ImageDirectory;
			});
			this.imageDirectories = imageDir;
			let ImageNumber = imageDir.length;
			let el = null;
			for(let i = 0; i <=ImageNumber-1;i++)
			{
				el = imageDir[i];
				el.setAttribute('draggable', 'true');
				 
				addEvent(el, 'dragstart', function (e) {
					console.log('dragstart for image' + this.id);

					e.dataTransfer.effectAllowed = 'copyLink';
					console.log('setting data: ' + this.id);

					e.dataTransfer.setData('Text', this.id); // required otherwise doesn't work
				});
				addEvent(el, 'drag', function (e) {
					console.log('drag');
				});

				addEvent(el, 'dragend', function (e) {
					console.log('dragend with effect: ' + e.dataTransfer.dropEffect);
				});
				addEvent(el, 'click', function (e) {
					console.log('Image ' + this.id +' clicked on draggable');
					console.log('click offsetX, offsetY of image' + this.id + ': ' + e.offsetX + ", " + e.offsetY);
				});

				// Custom event fires only if holdToDrag is true - critical to allow drag drop on mobile device. 
				addEvent(el, 'dnd-poly-dragstart-pending', function (e) {
					console.log('dnd-poly-dragstart-pending');
					e.preventDefault();
					e.stopPropagation(); // stop it here to prevent it bubble up
				});

				addEvent(el, 'dnd-poly-dragstart-cancel', function (e) {
					console.log('dnd-poly-dragstart-cancel');
					e.preventDefault();
					e.stopPropagation(); // stop it here to prevent it bubble up
				});
				
				 
			}
			
		});
	}
```

How can I make it work? Please help me.

**Comment by reppners** [comment]
From looking at your code you are referring to Angular which is distinctively different from AngularJS.

For Angular append this code inside your polyfill.ts
```ts
import { polyfill } from "mobile-drag-drop";
// optional import of scroll behaviour
import { scrollBehaviourDragImageTranslateOverride } from "mobile-drag-drop/scroll-behaviour";

// options are optional ;)
polyfill( {
  // use this to make use of the scroll behaviour
  dragImageTranslateOverride: scrollBehaviourDragImageTranslateOverride
} );

window.addEventListener( "touchmove", function() {
  // workaround to make scroll prevent work in iOS Safari > 10
} );
```

**Comment by reppners** [comment]
Closing this as solved, if you need more help just comment in this issue :)
